packages/asset-swapper/src/utils/aggregate.ts

```diff
@@ -0,0 +1,780 @@
+ */
+export const BUY_SOURCES = [ERC20BridgeSource.Uniswap, ERC20BridgeSource.Eth2Dai];
+
+/**
+ * Represents a node on a fill path.
+ */
+export interface Fill {
+    flags: number;
+    exclusionMask: number;
+    input: BigNumber;
+    output: BigNumber;
+    parent?: Fill;
+    data?: any;
+}
+
+// Internal `data` field for `Fill` objects.
+interface FillData {
+    source: ERC20BridgeSource;
+}
+
+// `FillData` for native fills.
+interface NativeFillData extends FillData {
+    order: OrderWithoutDomain;
+    orderInfo: OrderInfo;
+}
+
+/**
+ * Options for `improveMarketBuyAsync()` and `improveMarketSellAsync()`.
+ */
+export interface ImproveOrdersOpts {
+    /**
+     * Liquidity sources to exclude. Default is none.
+     */
+    excludedSources: ERC20BridgeSource[];
+    /**
+     * Whether to prevent mixing Kyber orders with Uniswap and Eth2Dai orders.
+     */
+    noConflicts: boolean;
+    /**
+     * Complexity limit on the search algorithm, i.e., maximum number of
+     * nodes to visit. Default is 1024.
+     */
+    runLimit: number;
+    /**
+     * When generating bridge orders, we use
+     * sampled rate * (1 - bridgeSlippage)
+     * as the rate for calculating maker/taker asset amounts.
+     * This should be a small positive number (e.g., 0.0005) to make up for
+     * small discrepancies between samples and truth.
+     * Default is 0.0005 (5 basis points).
+     */
+    bridgeSlippage: number;
+    /**
+     * Number of samples to take for each DEX quote.
+     */
+    numSamples: number;
+    /**
+     * Dust amount, as a fraction of the fill amount.
+     * Default is 0.01 (100 basis points).
+     */
+    dustThreshold: number;
+    /**
+     * A Web3 provider. Required.
+     */
+    provider: SupportedProvider;
+}
+
+const IMPROVE_ORDERS_OPTS_DEFAULTS = {
+    runLimit: 1024,
+    excludedSources: [],
+    provider: undefined,
+    bridgeSlippage: 0.0005,
+    dustThreshold: 0.01,
+    numSamples: 8,
+    noConflicts: false,
+};
+
+// Used internally by `FillsOptimizer`.
+interface FillsOptimizerContext {
+    currentPath: Fill[];
+    currentPathInput: BigNumber;
+    currentPathOutput: BigNumber;
+    currentPathFlags: number;
+}
+
+/**
+ * Represents an individual DEX sample from the sampler contract.
+ */
+export interface DexSample {
+    source: ERC20BridgeSource;
+    input: BigNumber;
+    output: BigNumber;
+}
+
+/**
+ * Flags for `Fill` objects.
+ */
+export enum FillFlags {
+    SourceNative = 0x1,
+    SourceUniswap = 0x2,
+    SourceEth2Dai = 0x4,
+    SourceKyber = 0x8,
+}
+
+/**
+ * Improve a market sell operation by (potentially) merging native orders with
+ * generated bridge orders.
+ * @param nativeOrders Native orders.
+ * @param takerAmount Amount of taker asset to sell.
+ * @param opts Options object.
+ * @return Improved orders.
+ */
+export async function improveMarketSellAsync(
+    nativeOrders: OrderWithoutDomain[],
+    takerAmount: BigNumber,
+    opts?: Partial<ImproveOrdersOpts>,
```

Comment:
`provider` is required but `opts` is optional. I'd suggest pulling out the provider explicitly.